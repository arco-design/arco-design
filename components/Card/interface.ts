import { CSSProperties, ReactNode, HTMLAttributes } from 'react';
import { Omit } from '../_util/type';

/**
 * @title Card
 */
export interface CardProps extends Omit<HTMLAttributes<HTMLDivElement>, 'title' | 'className'> {
  style?: CSSProperties;
  className?: string | string[];
  children?: ReactNode;
  id?: string;
  /**
   * @zh 是否有边框
   * @en Whether to render the border
   * @defaultValue true
   */
  bordered?: boolean;
  /**
   * @zh 是否为加载中
   * @en Whether on loading state
   */
  loading?: boolean;
  /**
   * @zh 是否可悬浮
   * @en Whether the card can be hovered
   */
  hoverable?: boolean;
  /**
   * @zh 卡片尺寸
   * @en Size of the card
   * @defaultValue default
   */
  size?: 'default' | 'small';
  /**
   * @zh 卡片标题
   * @en Title of the card
   */
  title?: string | ReactNode;
  /**
   * @zh 卡片右上角的操作区域
   * @en Content to render in the top-right corner of the card
   */
  extra?: string | ReactNode;
  /**
   * @zh 卡片封面
   * @en Cover of card
   */
  cover?: ReactNode;
  /**
   * @zh 卡片底部的操作组
   * @en The action list which shows at the bottom of the card
   */
  actions?: ReactNode[];
  /**
   * @zh 自定义标题区域样式
   * @en The additional css style to apply to card head
   */
  headerStyle?: CSSProperties;
  /**
   * @zh 内容区域自定义样式
   * @en The additional css style to apply to card content
   */
  bodyStyle?: CSSProperties;
}

/**
 * @title Card.Meta
 */
export interface CardMetaProps {
  style?: CSSProperties;
  className?: string | string[];
  actionList?: ReactNode[];
  /**
   * @zh 头像
   * @en Avatar of the card
   */
  avatar?: ReactNode;
  /**
   * @zh 标题
   * @en Title of the card
   */
  title?: string | ReactNode;
  /**
   * @zh 描述
   * @en Description of the card
   */
  description?: string | ReactNode;
}

/**
 * @title Card.Grid
 */
export interface CardGridProps {
  children?: ReactNode;
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 是否可以悬浮
   * @en Whether can be hovered
   */
  hoverable?: boolean;
}
