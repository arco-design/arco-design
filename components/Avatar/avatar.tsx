import React, {
  useImperativeHandle,
  useRef,
  useEffect,
  useContext,
  CSSProperties,
  PropsWithChildren,
  forwardRef,
} from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import AvatarContext from './context';
import { AvatarProps } from './interface';
import { isNumber } from '../_util/is';

const defaultProps: AvatarProps = {
  shape: 'circle',
  autoFixFontSize: true,
  triggerType: 'button',
};

const Avatar = forwardRef<HTMLElement, PropsWithChildren<AvatarProps>>(
  (props: PropsWithChildren<AvatarProps>, ref) => {
    const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);

    const prefixCls = getPrefixCls('avatar');
    const contextProps = useContext(AvatarContext);
    const mergedProps = { ...defaultProps, ...componentConfig?.Avatar, ...contextProps, ...props };

    const {
      className,
      shape,
      size,
      children,
      autoFixFontSize,
      triggerIcon,
      triggerIconStyle,
      triggerType,
      onClick,
      ...rest
    } = mergedProps;
    const style = { ...contextProps.style, ...props.style };

    const textRef = useRef(null);
    const avatarRef = useRef(null);

    useEffect(() => {
      if (autoFixFontSize) {
        autoFixFontSizeHandler();
      }
    }, [size, children]);

    useImperativeHandle(ref, () => avatarRef.current);

    // auto adjust font size
    function autoFixFontSizeHandler() {
      if (textRef.current) {
        const textWidth = textRef.current.clientWidth;
        const size = props.size || avatarRef.current.offsetWidth;
        const scale = size / (textWidth + 8);

        if (size && scale < 1) {
          textRef.current.style.transform = `scale(${scale}) translateX(-50%)`;
        }
      }
    }

    const classNames = cs(
      prefixCls,
      `${prefixCls}-${shape}`,
      {
        [`${prefixCls}-with-trigger-icon`]: triggerIcon,
        [`${prefixCls}-rtl`]: rtl,
      },
      className
    );
    const childrenList = React.Children.toArray(children);
    const isImage =
      childrenList.length === 1 &&
      React.isValidElement(childrenList[0]) &&
      (childrenList[0].type === 'img' || childrenList[0].type === 'picture');

    const _triggerIconStyle: CSSProperties = triggerIconStyle || {};
    if (
      triggerType === 'button' &&
      (!triggerIconStyle || (triggerIconStyle && !triggerIconStyle.color)) &&
      style &&
      style.backgroundColor
    ) {
      _triggerIconStyle.color = style.backgroundColor;
    }

    return (
      <div
        ref={avatarRef}
        onClick={onClick}
        {...rest}
        style={{
          width: size,
          height: size,
          fontSize: isNumber(size) ? size / 2 : '',
          ...style,
        }}
        className={classNames}
      >
        {isImage ? <span className={`${prefixCls}-image`}>{children}</span> : null}
        {!isImage && (
          <span ref={textRef} className={`${prefixCls}-text`}>
            {children}
          </span>
        )}
        {triggerIcon && (
          <div className={`${prefixCls}-trigger-icon-${triggerType}`} style={_triggerIconStyle}>
            {triggerIcon}
          </div>
        )}
      </div>
    );
  }
);

Avatar.displayName = 'Avatar';

export default Avatar;
