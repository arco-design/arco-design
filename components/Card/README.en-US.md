`````
Component / Data Display

# Card

Card is generally used as a container for concise introduction or information.
`````

%%Content%%

## API

### Card

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|bordered|Whether to render the border|boolean |`true`|
|hoverable|Whether the card can be hovered|boolean |`-`|
|loading|Whether on loading state|boolean |`-`|
|size|Size of the card|'default' \| 'small' |`default`|
|cover|Cover of card|ReactNode |`-`|
|extra|Content to render in the top-right corner of the card|string \| ReactNode |`-`|
|title|Title of the card|string \| ReactNode |`-`|
|actions|The action list which shows at the bottom of the card|ReactNode[] |`-`|
|bodyStyle|The additional css style to apply to card content|CSSProperties |`-`|
|className|Additional css class|string \| string[] |`-`|
|headerStyle|The additional css style to apply to card head|CSSProperties |`-`|
|style|Additional style|CSSProperties |`-`|

### Card.Meta

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|avatar|Avatar of the card|ReactNode |`-`|
|description|Description of the card|string \| ReactNode |`-`|
|title|Title of the card|string \| ReactNode |`-`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|

### Card.Grid

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|hoverable|Whether can be hovered|boolean |`-`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
