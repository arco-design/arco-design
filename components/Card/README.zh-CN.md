`````
组件 / 数据展示

# 卡片 Card

将信息分类后分标题、详情等区域聚合展现，一般作为简洁介绍或者信息的大盘和入口。
`````

%%Content%%

## API

### Card

|参数名|描述|类型|默认值|
|---|---|---|---|
|bordered|是否有边框|boolean |`true`|
|hoverable|是否可悬浮|boolean |`-`|
|loading|是否为加载中|boolean |`-`|
|size|卡片尺寸|'default' \| 'small' |`default`|
|cover|卡片封面|ReactNode |`-`|
|extra|卡片右上角的操作区域|string \| ReactNode |`-`|
|title|卡片标题|string \| ReactNode |`-`|
|actions|卡片底部的操作组|ReactNode[] |`-`|
|bodyStyle|内容区域自定义样式|CSSProperties |`-`|
|className|节点类名|string \| string[] |`-`|
|headerStyle|自定义标题区域样式|CSSProperties |`-`|
|style|节点样式|CSSProperties |`-`|

### Card.Meta

|参数名|描述|类型|默认值|
|---|---|---|---|
|avatar|头像|ReactNode |`-`|
|description|描述|string \| ReactNode |`-`|
|title|标题|string \| ReactNode |`-`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|

### Card.Grid

|参数名|描述|类型|默认值|
|---|---|---|---|
|hoverable|是否可以悬浮|boolean |`-`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
