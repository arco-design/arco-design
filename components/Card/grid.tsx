import React, { useContext } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import type { CardGridProps } from './interface';

function Grid(props: CardGridProps, ref) {
  const { children, style, className, hoverable } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('card-grid');
  return (
    <div
      ref={ref}
      style={style}
      className={cs(prefixCls, { [`${prefixCls}-hoverable`]: hoverable }, className)}
    >
      {children}
    </div>
  );
}

const CardComponent = React.forwardRef<unknown, CardGridProps>(Grid);

CardComponent.displayName = 'CardGrid';

export default CardComponent;

export { CardGridProps };
