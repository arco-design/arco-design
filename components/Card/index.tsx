import React, { useContext } from 'react';
import cs from '../_util/classNames';
import Meta from './meta';
import Grid from './grid';
import { ConfigContext } from '../ConfigProvider';
import Spin from '../Spin';
import { CardProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: CardProps = {
  size: 'default',
  bordered: true,
};

function Card(baseProps: CardProps, ref) {
  const { getPrefixCls, loadingElement, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<CardProps>(baseProps, defaultProps, componentConfig?.Card);
  const {
    className,
    children,
    bordered,
    loading,
    hoverable,
    size,
    title,
    extra,
    cover,
    actions,
    headerStyle,
    bodyStyle,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('card');

  const actionList =
    actions && actions.length ? (
      <div className={`${prefixCls}-actions`}>
        <div className={`${prefixCls}-actions-right`}>
          {actions.map((action, index) => (
            <span key={`action-${index}`} className={`${prefixCls}-actions-item`}>
              {action}
            </span>
          ))}
        </div>
      </div>
    ) : null;

  let isContainGrid = false;
  let isContainMeta = false;

  const handledChildren = React.Children.map(children, (element: JSX.Element) => {
    if (element && element.type) {
      if (element.type === Grid) {
        isContainGrid = true;
      } else if (element.type === Meta) {
        isContainMeta = true;
        return React.cloneElement(element, { actionList });
      }
    }
    return element;
  });

  return (
    <div
      {...rest}
      ref={ref}
      className={cs(
        prefixCls,
        `${prefixCls}-size-${size}`,
        {
          [`${prefixCls}-loading`]: loading,
          [`${prefixCls}-bordered`]: bordered,
          [`${prefixCls}-hoverable`]: hoverable,
          [`${prefixCls}-contain-grid`]: isContainGrid,
          [`${prefixCls}-rtl`]: rtl,
        },
        className
      )}
    >
      {title || extra ? (
        <div
          className={cs(`${prefixCls}-header`, { [`${prefixCls}-header-no-title`]: !title })}
          style={headerStyle}
        >
          {title && <div className={`${prefixCls}-header-title`}>{title}</div>}
          {extra && <div className={`${prefixCls}-header-extra`}>{extra}</div>}
        </div>
      ) : null}

      {cover ? <div className={`${prefixCls}-cover`}>{cover}</div> : null}

      <div className={`${prefixCls}-body`} style={bodyStyle}>
        {loading ? loadingElement || <Spin /> : handledChildren}
        {isContainMeta ? null : actionList}
      </div>
    </div>
  );
}

const ForwardRefCard = React.forwardRef<unknown, CardProps>(Card);

const CardComponent = ForwardRefCard as typeof ForwardRefCard & {
  Meta: typeof Meta;
  Grid: typeof Grid;
};

CardComponent.Meta = Meta;

CardComponent.Grid = Grid;

CardComponent.displayName = 'Card';

export default CardComponent;

export { CardProps };
