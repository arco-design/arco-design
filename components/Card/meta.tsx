import React, { useContext } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import type { CardMetaProps } from './interface';

function Meta(props: CardMetaProps, ref) {
  const { className, title, avatar, description, actionList, ...others } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('card-meta');
  const classNames = cs(prefixCls, className);

  return (
    <div {...others} ref={ref} className={classNames}>
      {title || description ? (
        <div className={`${prefixCls}-content`}>
          {title && <div className={`${prefixCls}-title`}>{title}</div>}
          {description && <div className={`${prefixCls}-description`}>{description}</div>}
        </div>
      ) : null}
      {avatar || actionList ? (
        <div
          className={cs(`${prefixCls}-footer `, { [`${prefixCls}-footer-only-actions`]: !avatar })}
        >
          {avatar ? <div className={`${prefixCls}-avatar`}>{avatar}</div> : null}
          {actionList}
        </div>
      ) : null}
    </div>
  );
}

const MetaComponent = React.forwardRef<unknown, CardMetaProps>(Meta);

MetaComponent.displayName = 'CardMeta';

export default MetaComponent;

export { CardMetaProps };
