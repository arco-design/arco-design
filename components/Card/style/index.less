@import './token.less';
@import '../../style/mixin.less';

@card-prefix-cls: ~'@{prefix}-card';

.@{card-prefix-cls} {
  position: relative;
  background: @card-color-bg;
  transition: box-shadow @transition-duration-2 @transition-timing-function-linear;
  border-radius: @card-border-radius-no-border;

  &-header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: relative;
    box-sizing: border-box;
    border-bottom: @card-border-width-title-bottom solid @card-color-border;
    overflow: hidden;

    &-no-title::before {
      content: ' ';
      display: block;
    }

    &-title {
      flex: 1;
      font-weight: @card-font-weight-title;
      color: @card-color-title;
      .text-ellipsis();
    }

    &-extra {
      color: @card-color-title-extra;
      .text-ellipsis();
    }
  }

  &-body {
    color: @card-color-body;
  }

  &-cover {
    overflow: hidden;

    > * {
      display: block;
      width: 100%;
    }
  }

  &-actions {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: @card-margin-top-meta-footer;

    &::before {
      content: '';
      visibility: hidden;
    }

    &-right {
      display: flex;
      align-items: center;
    }

    &-item {
      display: flex;
      align-items: center;
      justify-content: center;
      color: @card-color-action;
      cursor: pointer;
      .text-ellipsis();

      transition: color @transition-duration-2 @transition-timing-function-linear;

      &:hover {
        color: @card-color-action_hover;
      }

      &:not(:last-child) {
        margin-right: @card-margin-right-action-item;
      }
    }
  }

  &-meta {
    &-footer {
      display: flex;
      align-items: center;
      justify-content: space-between;

      &:last-child {
        margin-top: @card-margin-top-meta-footer;
      }

      &-only-actions::before {
        content: '';
        visibility: hidden;
      }

      .@{card-prefix-cls}-actions {
        margin-top: 0;
      }
    }

    &-title {
      font-weight: @card-font-weight-title;
      color: @card-color-title;
      .text-ellipsis();
    }

    &-description:not(:first-child) {
      margin-top: @card-margin-top-meta-description;
    }
  }

  &-grid {
    position: relative;
    box-sizing: border-box;
    width: 33.33%;
    box-shadow: @card-border-width 0 0 0 @card-color-border,
      0 @card-border-width 0 0 @card-color-border,
      @card-border-width @card-border-width 0 0 @card-color-border,
      @card-border-width 0 0 0 @card-color-border inset,
      0 @card-border-width 0 0 @card-color-border inset;

    &::before {
      content: '';
      pointer-events: none;
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      transition: box-shadow @transition-duration-2 @transition-timing-function-linear;
    }

    &-hoverable:hover {
      z-index: 1;

      &::before {
        box-shadow: 0 4px 10px @card-color-box-shadow;
      }
    }

    // 避免 Card 的背景色遮住由 box-shadow 实现的 Grid 的边框
    .@{card-prefix-cls} {
      background: none;
      box-shadow: none;
    }
  }

  // 以下为特殊状态的补充
  &-contain-grid:not(&-loading) > &-body {
    display: flex;
    flex-wrap: wrap;
    margin: 0 -@card-border-width;
    padding: 0;
  }

  &-hoverable:hover {
    box-shadow: 0 4px 10px @card-color-box-shadow;
  }

  &-bordered {
    border: @card-border-width solid @card-color-border;
    border-radius: @card-border-radius;

    .@{card-prefix-cls}-cover {
      border-radius: @card-border-radius @card-border-radius 0 0;
    }
  }

  &-loading &-body {
    overflow: hidden;
    text-align: center;
  }

  // 不同尺寸
  .size(@size) {
    &-size-@{size} {
      font-size: ~'@{card-size-@{size}-font-size}';

      .@{card-prefix-cls}-header {
        height: ~'@{card-size-@{size}-height-title}';
        padding-left: ~'@{card-size-@{size}-padding-horizontal-title}';
        padding-right: ~'@{card-size-@{size}-padding-horizontal-title}';
      }

      .@{card-prefix-cls}-header-title,
      .@{card-prefix-cls}-meta-title {
        font-size: ~'@{card-size-@{size}-font-size-title}';
      }

      .@{card-prefix-cls}-header-extra {
        font-size: ~'@{card-size-@{size}-font-size-title-extra}';
      }

      .@{card-prefix-cls}-body {
        padding: ~'@{card-size-@{size}-padding-vertical-body}' ~'@{card-size-@{size}-padding-horizontal-body}';
      }
    }
  }

  .size(default);
  .size(small);

  &-rtl &-actions-item:not(:last-child) {
    margin-left: @card-margin-right-action-item;
    margin-right: 0;
  }
}

// dark mode

@{arco-theme-tag} {
  &[arco-theme='dark'] {
    .@{card-prefix-cls}-grid-hoverable:hover::before,
    .@{card-prefix-cls}-hoverable:hover {
      box-shadow: 0 4px 10px @card-color-box-shadow_dark;
    }
  }
}
