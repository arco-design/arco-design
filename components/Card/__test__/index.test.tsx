import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Card from '..';
import Avatar from '../../Avatar';

const { Meta } = Card;

mountTest(Card);
componentConfigTest(Card, 'Card');

function mountCard(component: React.ReactElement) {
  return mount(component);
}

function createCard() {
  return (
    <Card title="Default Card">
      <p>Content</p>
      <p>Content</p>
      <p>Content</p>
    </Card>
  );
}

function createMeta() {
  return (
    <Card>
      <Meta
        avatar={<Avatar style={{ marginRight: 20, backgroundColor: '#5babf3' }}>Docor</Avatar>}
        title="Docor"
        description="This is Docor"
      />
    </Card>
  );
}

describe('Card', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('Card display correctly', () => {
    const wrapper = mountCard(createCard());
    expect(wrapper.props().title).toBe('Default Card');
  });

  it('Meta display correctly', () => {
    const wrapper = mountCard(createMeta());
    expect(wrapper.find('.arco-card-meta-title').at(0).text()).toBe('Docor');
  });
});
