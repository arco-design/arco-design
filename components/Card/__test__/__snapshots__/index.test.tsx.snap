// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Card default 1`] = `
<div
  class="arco-card arco-card-size-default arco-card-bordered"
>
  <div
    class="arco-card-body"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Card set className globally 1`] = `
<div
  class="arco-card arco-card-size-default arco-card-bordered global-default-class"
>
  <div
    class="arco-card-body"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Card set className globally and component set className itself. 1`] = `
<div
  class="arco-card arco-card-size-default arco-card-bordered component-class"
>
  <div
    class="arco-card-body"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Card set data-* property 1`] = `
<div
  class="arco-card arco-card-size-default arco-card-bordered"
  data-test-name="Card"
>
  <div
    class="arco-card-body"
  />
</div>
`;
